### PR TITLE
feat(longtasks): allow callback to add span attributes on collection

### PR DESCRIPTION
**Pull request by crellison**
## Which problem is this PR solving?

- ability to add application-specific context as attributes to `longtask` spans

## Short description of the changes

This adds a callback parameter to the LongTaskInstrumentation to be executed when a longtask is observed. The return value of the callback is an object of span attributes to be attached to the longtask span. 

This feature is important for web applications, as additional details on application state may be relevant to long tasks and can help in correlating long tasks with slow code. This feature allows sites using client-side navigation to add window location details to spans and enables hydrating spans with relevant information from application state.

## Checklist

- [x] Ran `npm run test-all-versions` for the edited package(s) on the latest commit if applicable. (not applicable, but checking it rather than removing)

### CODE REVIEW

File: plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts
Comment (blumamir) [inline plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts:55]: I am not very familiar with the web instrumentations, but in node we usually add this kind of hook in the `config` object.
Does it make sense to do that for web instrumentations as well?

File: plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts
Comment (blumamir) [inline plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts:81]: Should we have protection for the case this callback throws? (try / catch)
This is an external callback that the instrumentation has no control of.
in node we usually wrap these kinds of calls with `safeExecuteInTheMiddle`

File: plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts
Comment (blumamir) [inline plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts:37]: a common practice for this repo is to have a `types.ts` file and have all types declarations exported from there.
[example](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/plugins/web/opentelemetry-instrumentation-user-interaction/src/types.ts)
I suggest to use it here as well

File: plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts
Comment (crellison) [reply plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts]: Totally makes sense.

File: plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts
Comment (crellison) [reply plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts]: Definitely safer with a try-catch. Added the feature and a test for that behavior

File: plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts
Comment (blumamir) [inline plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts:34]: you don't have to store in as a private member.
the config object is accessible via `this._config.observerCallback` where you need to use it

File: plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts
Comment (blumamir) [inline plugins/web/opentelemetry-instrumentation-long-task/src/instrumentation.ts:70]: We usually have this style for logging these kinds of errors:
`diag.error('longtask instrumentation: observer hook failed', err);`

File: plugins/web/opentelemetry-instrumentation-long-task/src/types.ts
Comment (blumamir) [inline plugins/web/opentelemetry-instrumentation-long-task/src/types.ts:20]: Can you please write a small comment to document this callback for potential users?
We usually also add it to the README. you can see examples in other instrumentations such as [aws-sdk](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/opentelemetry-instrumentation-aws-sdk#aws-sdk-instrumentation-options)

File: plugins/web/opentelemetry-instrumentation-long-task/src/types.ts
Comment (blumamir) [inline plugins/web/opentelemetry-instrumentation-long-task/src/types.ts:16]: in other instrumentations, we usually pass the `span` directly as a function parameter and allow the user to `setAttributes` on it directly.
It's a bit more flexible as the user can attach more than just attributes (events), and can also derive logic based on the span (for example, if this span is not recording, skip executing heavy logic)

File: plugins/web/opentelemetry-instrumentation-long-task/test/longTask.test.ts
Comment (blumamir) [inline plugins/web/opentelemetry-instrumentation-long-task/test/longTask.test.ts:114]: This hook should be updated to the new signature (receive span as function argument and store attributes from the hook code)

File: plugins/web/opentelemetry-instrumentation-long-task/src/types.ts
Comment (blumamir) [inline plugins/web/opentelemetry-instrumentation-long-task/src/types.ts:33]: Thanks for addressing all the comments :)
One last request: some hook functions started small and got bigger over time as more data was added to the function signature as arguments.
We usually try to avoid that by sending all the hook data as `info` object which can be extended in the future in a non-breaking fashion and keeps the function signature and implementation short and lean.
You can see example of that in `aws-sdk` [here](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/6ba387ab45d0f67fdfac85c43bb0f0f67f2a119b/plugins/node/opentelemetry-instrumentation-aws-sdk/src/types.ts#L44).

**Comment by blumamir** [comment]
Thanks @crellison for addressing everything 💯 
This is great work.
Added last minor comment and then it can be merged as far as I am concerned

File: plugins/web/opentelemetry-instrumentation-long-task/src/types.ts
Comment (crellison) [reply plugins/web/opentelemetry-instrumentation-long-task/src/types.ts]: That's a good point. Thanks for all the feedback on this! We're excited to see it in prod!

**Review by blumamir — APPROVED** [review]
LGTM.
Thanks for the contribution :)

I'll be happy for another review from someone with stronger web instrumentation background
@open-telemetry/javascript-approvers

**Comment by crellison** [comment]
@blumamir Do you have a way of checking in with any of the JS approvers to see if someone is available to review?

**Comment by blumamir** [comment]
> @blumamir Do you have a way of checking in with any of the JS approvers to see if someone is available to review?

I suggest writing in the `opentelemetry-js` CNCF slack channel.
@dyladan - Who from the approvers is most familiar with the web instrumentations?

**Review by vmarchaud — APPROVED** [review]

**Review by t2t2 — APPROVED** [review]

**Comment by Rauno56** [comment]
@t2t2, the component owner, has approved, I think we are good to go, unless @crellison wants to address the nit above.

**Review by mwear — APPROVED** [review]
LGTM

File: plugins/web/opentelemetry-instrumentation-long-task/test/longTask.test.ts
Comment (crellison) [reply plugins/web/opentelemetry-instrumentation-long-task/test/longTask.test.ts]: Fixed assert strings

**Comment by dyladan** [comment]
> @dyladan - Who from the approvers is most familiar with the web instrumentations?

Probably @MSNev at this point. @vmarchaud has more web experience than me. Bart was the person who did most of our web reviews before.

**Review by MSNev — APPROVED** [review]

**Review by dyladan — APPROVED** [review]

**Pull request merged**
